### PR TITLE
feat: create a new EditableHeading component (with new typography styles)

### PR DESCRIPTION
**Pull request by talkor**
Screenshot 2023-10-31 at 19 58 25
https://monday.monday.com/boards/3532714909/pulses/5266220954

### CODE REVIEW

File: src/components/EditableHeading/HiddenInputPlaceholder.module.scss [hunk 1]
```diff
@@ -0,0 +1,7 @@
+.hiddenInputPlaceholder {
+  position: absolute;
```
Comment (talkor) [inline src/components/EditableHeading/HiddenInputPlaceholder.module.scss:2]: I can't use the `hidden-element` mixin here because I need the width

File: src/components/EditableHeading/__stories__/EditableHeading.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  </Story>
+</Canvas>
+
+## Props
+
+<ArgsTable story="Overview" />
+
+## Usage
+
+<UsageGuidelines guidelines={[]} />
```
Comment (talkor) [inline src/components/EditableHeading/__stories__/EditableHeading.stories.mdx:54]: The old guidelines are not relevant anymore, I'll talk to Meirav

File: src/components/EditableHeading/EditableHeading.tsx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      <div
+        ref={mergedRef}
+        id={id}
+        aria-label={ariaLabel}
+        data-testid={dataTestId || getTestId(ComponentDefaultTestId.EDITABLE_HEADING, id)}
+        className={cx(styles.editableHeading, className)}
+      >
+        {isEditing && editable ? (
+          <>
+            <HiddenInputPlaceholder className={inputClassNames} value={inputValue} onChange={setInputWidth} />
```
Comment (talkor) [inline src/components/EditableHeading/EditableHeading.tsx:137]: The main difference between this and the old one is that the input is now an inline element and not a black element, so since html inputs cannot have a dynamic width that fits the text we use this hack of a hidden placeholder that defines the width for the input. This is a very common hack and it's being used in libraries such as `react-input-autosize`.

File: src/components/EditableHeading/EditableHeading.module.scss [hunk 4]
```diff
@@ -0,0 +1,36 @@
+@import "../../styles/typography";
+@import "~monday-ui-style/dist/mixins";
+
+@include create-title-classes;
+
+.editableHeading {
+  @include ellipsis;
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.module.scss:7]: why is this needed? aren't you getting it from the `Heading` component?

File: src/components/EditableHeading/EditableHeading.tsx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+export interface EditableHeadingProps extends VibeComponentProps {
+  value: string;
+  type?: HeadingType;
+  weight?: HeadingWeight;
+  editable?: boolean;
+  onChange?: (value: string) => void;
+  focusOnMount?: boolean;
+  selectOnMount?: boolean;
+  ariaLabel?: string;
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.tsx:24]: add description per prop for its documenation

File: src/components/EditableHeading/EditableHeading.tsx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    {
+      id,
+      className,
+      "data-testid": dataTestId,
+      value,
+      type = Heading.types.H1,
+      weight = HeadingWeight.NORMAL,
+      editable = true,
+      onChange = () => {},
+      focusOnMount = true,
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.tsx:41]: should probably be `false`

File: src/components/EditableHeading/EditableHeading.tsx [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  (
+    {
+      id,
+      className,
+      "data-testid": dataTestId,
+      value,
+      type = Heading.types.H1,
+      weight = HeadingWeight.NORMAL,
+      editable = true,
+      onChange = () => {},
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.tsx:40]: I guess you can keep it unassigned

File: src/components/EditableHeading/EditableHeading.tsx [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+} = forwardRef(
+  (
+    {
+      id,
+      className,
+      "data-testid": dataTestId,
+      value,
+      type = Heading.types.H1,
+      weight = HeadingWeight.NORMAL,
+      editable = true,
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.tsx:39]: I usually prefer props that are "off" by default. meaning that if I want to turn something on, I activate it instead of making something inactive

for example, prefer:
`<EditableHeading readOnly/>`
instead of
`<EditableHeading editable={false}/>`

File: src/components/EditableHeading/EditableHeading.tsx [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        onClick: toggleEditMode,
+        id,
+        role: "button"
+      },
+      mergedRef
+    );
+
+    function focus() {
+      if (inputRef.current) {
+        requestAnimationFrame(() => {
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.tsx:102]: why is this needed?

File: src/components/EditableHeading/EditableHeading.tsx [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    useLayoutEffect(() => {
+      if (focusOnMount) {
+        focus();
+      }
+      if (selectOnMount) {
+        select();
+      }
+    }, [focusOnMount, isEditing, selectOnMount]);
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.tsx:121]: this means the hook will run even if `focusOnMount` changes, which isn't the expected behavior.
You might want to split it into two effects, one that runs once on mounting, and another only when the `isEditing` changes

File: src/components/EditableHeading/EditableHeading.module.scss [hunk 11]
```diff
@@ -0,0 +1,36 @@
+@import "../../styles/typography";
+@import "~monday-ui-style/dist/mixins";
+
+@include create-title-classes;
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.module.scss:4]: what is this exactly?

File: src/components/Modal/__stories__/modal.stories.mdx [hunk 12]
```diff
@@ -410,7 +410,7 @@ clicking outside. Use this variant in case of sensitive or important messages, a
             {/** Please follow the design guidelines when implementing custom title **/}
             <ModalHeader
               description={"Description text goes here"}
-              title={<LegacyEditableHeading type={LegacyEditableHeading.types.h2} placeholder={"Modal title"} />}
+              title={<EditableHeading type={EditableHeading.types.H2} value={"Modal title"} />}
```
Comment (shlomitc) [inline src/components/Modal/__stories__/modal.stories.mdx:413]: it behaves differently from the previous component. It should span to 100%

now:
image

before:
image

File: src/components/EditableHeading/EditableHeading.module.scss [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+.editableHeading {
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.module.scss:6]: I took this example from the modal with the editable header. 
It should be aligned the same as the description text, and when hovered, the border should be applied outside this line. you can talk with Meirav if it is not clear

image

**Review by shlomitc — CHANGES_REQUESTED** [review]
once PR #1726 is merged - please add also the deprecation to the legacy component, the same as you can see in the `LegacyHeading` page
image

File: src/components/EditableHeading/EditableHeading.module.scss (hunk 11 above)
Comment (talkor) [reply src/components/EditableHeading/EditableHeading.module.scss:4]: This creates the new typography classes

File: src/components/Modal/__stories__/modal.stories.mdx (hunk 12 above)
Comment (talkor) [reply src/components/Modal/__stories__/modal.stories.mdx:413]: That's intentional, the new spec defines it as an inline element
Screenshot 2023-11-01 at 12 22 26

See https://github.com/mondaycom/monday-ui-react-core/pull/1727#discussion_r1378548622

File: src/components/EditableHeading/EditableHeading.module.scss (hunk 13 above)
Comment (talkor) [reply src/components/EditableHeading/EditableHeading.module.scss:6]: I'll make sure with Meirav, the old component behaves the same, if we shift it to the left it means the the input will overflow

File: src/components/EditableHeading/EditableHeading.tsx (hunk 6 above)
Comment (talkor) [reply src/components/EditableHeading/EditableHeading.tsx:41]: That's the default in the old component, I'll make sure with Meirav though

File: src/components/EditableHeading/EditableHeading.module.scss (hunk 11 above)
Comment (shlomitc) [reply src/components/EditableHeading/EditableHeading.module.scss:4]: Ok, I see. It’s for the input element. Just add a comment there to make it more clear as it looks like a side affect that no one uses. Maybe even using it inside some class that you use to explain the connecting

File: src/components/Modal/__stories__/modal.stories.mdx (hunk 12 above)
Comment (shlomitc) [reply src/components/Modal/__stories__/modal.stories.mdx:413]: As long as it’s verified with Meirav, we’re good

File: src/components/EditableHeading/EditableHeading.module.scss (hunk 13 above)
Comment (shlomitc) [reply src/components/EditableHeading/EditableHeading.module.scss:6]: The input should probably contain some padding so the text inside would be aligned, but the input frame will overflow

Comment (talkor) [reply src/components/EditableHeading/EditableHeading.module.scss:6]: Yes but it might be a big deal to shift it in `overflow: hidden` containers like this one, we'll have to use a portal for this
Screenshot 2023-11-01 at 12 45 58

Comment (shlomitc) [reply src/components/EditableHeading/EditableHeading.module.scss:6]: Or we can solve this with some padding games. Give the content some padding and reduce it from the container

**Comment by talkor** [comment]
Accidentally deleted your comment about the focus hook :( Will resolve it

File: src/components/EditableHeading/EditableHeading.module.scss (hunk 13 above)
Comment (talkor) [reply src/components/EditableHeading/EditableHeading.module.scss:6]: Wouldn't that be too hacky? It means I'll have to adjust the `Modal*` components' width just for this case. And it would have to be adjusted as well in other overflowed containers

Comment (shlomitc) [reply src/components/EditableHeading/EditableHeading.module.scss:6]: Yes, it is an issue that will appear in any container that is overflowing, and that the content sticks to the side.
But I don't think it's that hacky... it's just how you split the padding. it might get complex if all the rendered elements (header, description, content) aren't wrapped with any wrapper from the component.

But let's give it a shot and see if it goes bad or not

File: src/components/EditableHeading/EditableHeading.tsx [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+import useClickableProps from "../../hooks/useClickableProps/useClickableProps";
+
+export interface EditableHeadingProps extends VibeComponentProps {
+  /** Heading text */
+  value: string;
+  /** Heading type */
+  type?: HeadingType;
+  /** Heading weight */
+  weight?: HeadingWeight;
+  /** onChange callback */
```
Comment (shlomitc) [inline src/components/EditableHeading/EditableHeading.tsx:23]: I meant here to something that explains better what it does. for example in this case, `will be called whenever the current value changes to a non-empty value, `

File: src/components/EditableHeading/EditableHeading.tsx (hunk 6 above)
Comment (shlomitc) [reply src/components/EditableHeading/EditableHeading.tsx:49]: It's more of a business-decision, so I'd suggest keeping the component non-opinionated and allow the consumer to decide if he wants to focus

File: src/components/EditableHeading/__tests__/EditableHeading-interactions.tsx [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+}
+function getHeading(canvas: Canvas) {
+  return getByTestId(canvas, ComponentDefaultTestId.CLICKABLE);
+}
+
+function getInput(canvas: Canvas) {
+  return getByRole(canvas, "input");
+}
+
+async function testBasicEditing(canvas: Canvas) {
```
Comment (shlomitc) [inline src/components/EditableHeading/__tests__/EditableHeading-interactions.tsx:28]: I try to name the tests by what they should do.
For example, `editAndChangeToValidText`

That forces me to write tests that are simple and doesn't mix multiple tests in one

File: src/components/EditableHeading/__tests__/EditableHeading-interactions.tsx [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  clickElement(compponent);
+
+  const input = getInput(canvas);
+  await clearText(input);
+  await typeText(input, text);
+  expect(input).toHaveAttribute("value", text);
+
+  await resetFocus();
+  const heading = getHeading(canvas);
+  expect(heading).toHaveTextContent(text);
```
Comment (shlomitc) [inline src/components/EditableHeading/__tests__/EditableHeading-interactions.tsx:39]: worth splitting to another test, just so chromatic will catch both

**Review by shlomitc — CHANGES_REQUESTED** [review]

File: src/components/EditableHeading/EditableHeading.tsx (hunk 6 above)
Comment (talkor) [reply src/components/EditableHeading/EditableHeading.tsx:49]: Talked to Meirav about this and she said that editing the text is the purpose so it should be focused by default, and also she suggested to keep it the same as the old component's default

File: src/components/EditableHeading/__tests__/EditableHeading-interactions.tsx (hunk 16 above)
Comment (talkor) [reply src/components/EditableHeading/__tests__/EditableHeading-interactions.tsx:39]: What do you mean? I want to check that the text is changing when you blur after you change it, and the next test also checks the case where you just clear the text

File: src/components/EditableHeading/EditableHeading.tsx (hunk 6 above)
Comment (shlomitc) [reply src/components/EditableHeading/EditableHeading.tsx:49]: None of the html elements are focused by default. This is always an opt-in feature and I can think of product cases that doesn’t require editing by default (board header). So let’s avoid from making it on by default

File: src/components/EditableHeading/__tests__/EditableHeading-interactions.tsx (hunk 16 above)
Comment (shlomitc) [reply src/components/EditableHeading/__tests__/EditableHeading-interactions.tsx:39]: So the first text will be the blur only. The second test would be a component which already has a value set and then clear it

File: src/components/EditableHeading/EditableHeading.module.scss (hunk 13 above)
Comment (talkor) [reply src/components/EditableHeading/EditableHeading.module.scss:4]: I added an opt-out prop (`alignWithText` - if you have a better name LMK) that adds negative margin to the component so it will be aligned with the text. In case of hidden overflowing containers users can disable it. Specifically the `ModalHeader` has a different issue when providing a custom title but the might be a breaking change, I suggest discussing it further

File: src/components/EditableHeading/EditableHeading.tsx (hunk 6 above)
Comment (talkor) [reply src/components/EditableHeading/EditableHeading.tsx]: Just to be clear - the `onMount` refers to the editing mode, not the view mode. E.g. you can on the component, the input is shown, and only then you focus on the input

Comment (shlomitc) [reply src/components/EditableHeading/EditableHeading.tsx]: So why do you need to props. If it’s in edit mode - isn’t it always focused?

Comment (talkor) [reply src/components/EditableHeading/EditableHeading.tsx]: Which two props? I'm referring only to the focusOnMount prop, it can be disabled when you want to control the focus externally for example, or if you want to select the text on mount which is also a feature we provide

File: src/components/Modal/ModalHeader/ModalHeader.tsx [hunk 17]
```diff
@@ -82,19 +87,20 @@ const ModalHeader: FC<ModalHeaderProps> = ({
...
-        id={id}
-        maxLines={2}
-        data-testid={dataTestId || getTestId(ComponentDefaultTestId.MODAL_HEADER, id)}
-        className={titleClassName}
-      >
-        {icon && (
-          <span className={cx(styles.icon, iconClassName)}>
-            <Icon icon={icon} iconType={Icon.type.SVG} iconSize={iconSize} ignoreFocusStyle clickable={false} />
-          </span>
+      <div data-testid={dataTestId || getTestId(ComponentDefaultTestId.MODAL_HEADER, id)}>
```
Comment (shlomitc) [inline src/components/Modal/ModalHeader/ModalHeader.tsx:90]: can't you simply set the data-testid on the root div of the component? do we really need this extra div?

File: src/components/EditableHeading/__stories__/EditableHeading.stories.mdx [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+## Usage
+
+<UsageGuidelines
+  guidelines={[
+    "Use when the title (Heading) is editable and you want to edit it",
+    "The component takes 100% of width and height of its container"
+  ]}
+/>
+## Variants
```
Comment (shlomitc) [inline src/components/EditableHeading/__stories__/EditableHeading.stories.mdx:60]: probably needs an empty line before
image

File: src/components/EditableHeading/__stories__/EditableHeading.stories.mdx [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+## Props
+
+<ArgsTable story="Overview" />
+
+## Usage
+
+<UsageGuidelines
+  guidelines={[
+    "Use when the title (Heading) is editable and you want to edit it",
+    "The component takes 100% of width and height of its container"
```
Comment (shlomitc) [inline src/components/EditableHeading/__stories__/EditableHeading.stories.mdx:57]: is it? 
image

Comment (talkor) [reply src/components/EditableHeading/__stories__/EditableHeading.stories.mdx:57]: I agree but I asked Meirav specifically about this and she said that even though the header itself is an inline element, the entire component takes the entire block which is true

**Review by shlomitc — APPROVED** [review]
looks good
